Feature/controllers apps reqs task 2

osmanaliyardim opened the pull request:
### [Task 2](https://learn.epam.com/myLearning/program?groupGuid=68c6d8bc-9684-4ee5-a549-048b3ef481d8) - Action Results and File Uploads

### What this change does
> Focus on why the change was made and provide short, clean and good summary
* Adding functionality to show images on Categories index with details

### Code changes
> Explain what files and code parts were changed in scope of PR and reviewer should pay attention for
* Added new Category Add and Update views
* Updated Category index view
* Some code refactorings
* Category controller adjustments for file uploading

### What has been done to confirm PR works and does not break existing stuff
> Provide guarantees that your change does not ruin existing functionality and works as expected
* Tested on localhost.

### Screenshots
> Screenshots with descriptions confirming your correct results

* Current view of Categories/Index
![categories-index](https://user-images.githubusercontent.com/25593515/198511181-1e355461-0512-41ac-bd10-e7701134f137.png)

osmanaliyardim committed:
FEATURE: Image Upload

Image upload and edit functionality has been added for Categories.

osmanaliyardim committed:
FIX: Bugs

Some bugs have been solved.

osmanaliyardim committed:
REFACTOR: Category

Categories index page has been updated.

dotnet-fizzyy commented on Business/Abstract/ICategoryService.cs:
```diff
@@ -8,4 +8,7 @@ public interface ICategoryService
 
     Task<Category> GetByIdAsync(Guid categoryId);
 
+    Task AddAsync(Category category);
```
Just want to clarify: is it entity or DTO? If first option, please, replace it with DTO to avoid violations with architecture and making _Presentation_ layer depend on _Database_ layer

Same case for `IProductService`

dotnet-fizzyy commented on Core/DataAccess/EntityFramework/EfEntityRepositoryBase.cs:
```diff
@@ -44,12 +44,28 @@ public async Task<TEntity> GetAsync(Expression<Func<TEntity, bool>> filter)
+        await DbContext.SaveChangesAsync();
+    }
+
+    // ToDo: Exract it as an extension class
```
Well, I guess you may leave it in repo layer as well

dotnet-fizzyy commented on Core/DataAccess/EntityFramework/EfEntityRepositoryBase.cs:
```diff
@@ -44,12 +44,28 @@ public async Task<TEntity> GetAsync(Expression<Func<TEntity, bool>> filter)
+    }
+
+    // ToDo: Exract it as an extension class
+    public void DetachLocal<T>(DbContext context, T t, Guid entryId)
```
Please, give more clear name for `T` argument, not just `t` variable :)

dotnet-fizzyy commented on WebNetSample/Views/Categories/Index.cshtml:
```diff
@@ -2,12 +2,56 @@
+</p>
+
+<table class="table table-striped table-hover">
+
```
Guess this row is extra one

osmanaliyardim replied:
No, it is not actually. It is the start of html table element..

dotnet-fizzyy commented on WebNetSample/Controllers/CategoriesController.cs:
```diff
@@ -19,4 +20,51 @@ public async Task<IActionResult> Index()
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> Update(Category updatedCategory)
```
Please, move all logic related to updating your image to _Service_ layer.

Remember the **golden rule of controller**: it should **not** operate any logic (exception cases are only about some View models/HTTP processing). All logic related entity processing should be stored **only** on _Business_ layer.

dotnet-fizzyy commented on WebNetSample/Controllers/CategoriesController.cs:
```diff
@@ -19,4 +20,51 @@ public async Task<IActionResult> Index()
+    [HttpPost]
+    public async Task<IActionResult> Update(Category updatedCategory)
+    {
+        string FileName = Path.GetFileNameWithoutExtension(updatedCategory.ImageFile?.FileName);
```
1. Use `var` instead of explicit type declaration
2. If your `updatedCategory.ImageFile` in DTO model is empty I guess then it has no sense to leave all logic related to image processing later. I see an option where you can check for `null` value of `updatedCategory.ImageFile` and skip all file processing. Otherwise, if image is a demand, than null-guard operator (`?`) is not necessary

dotnet-fizzyy commented on WebNetSample/Controllers/CategoriesController.cs:
```diff
@@ -19,4 +20,51 @@ public async Task<IActionResult> Index()
+            updatedCategory.ImageFile?.CopyTo(fileStream);
+        }
+
+        updatedCategory.ImagePath = "\\Images\\" + FileName;
```
Let's put `\\Images\\` into a constant

dotnet-fizzyy commented on Entity/Concrete/Category.cs:
```diff
@@ -1,20 +1,39 @@
+    public string ImagePath { get; set; }
+
+    [NotMapped]
+    public IFormFile? ImageFile { get; set; }
```
No need to use `IFormFile` in entity, entity should store **only** the path to your file, not its content. Please, move `IFormFile` to the DTO

dotnet-fizzyy commented on Entity/Concrete/Product.cs:
```diff
@@ -8,16 +11,21 @@ public class Product : BaseEntity
+    public string ImagePath { get; set; }
+
+    [NotMapped]
+    public IFormFile? ImageFile { get; set; }
```
Same case with `IFormFile` for the entity

dotnet-fizzyy commented on Core/DataAccess/EntityFramework/EfEntityRepositoryBase.cs:
```diff
@@ -44,12 +44,28 @@ public async Task<TEntity> GetAsync(Expression<Func<TEntity, bool>> filter)
+        var local = context.Set<T>()
+            .Local
+            .FirstOrDefault(entry => entry.Id.Equals(entryId));
+        if (!(local == null))
```
```suggestion
        if (local != null)
```

dotnet-fizzyy commented on WebNetSample/Views/Categories/Index.cshtml:
```diff
@@ -2,12 +2,56 @@
+
+    <thead>
+        <tr>
+            <th colspan="7" class="text-center bg-dark text-light">CATEGORIES WITH DETAILS</th>
```
CSS trick: you can easily use [`text-transform`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform) CSS property to make you text format in every way you need, just keep in mind that ;)

dotnet-fizzyy commented on Entity/Concrete/ProductDetails.cs:
```diff
@@ -8,7 +9,8 @@ public class ProductDetails : BaseEntity
 
-    public string ImageUrl { get; set; }
+    public IFormFile? ImageFile { get; set; }
+    public string ImagePath { get; set; }
```
Don't forget to add empty row between properties as a separator, please

osmanaliyardim replied:
Can you [this](https://github.com/osmanaliyardim/WebNetSample/pull/7#:~:text=%E2%80%A6-,dc9b231,-Merge%20state) commit?

osmanaliyardim committed:
FEATURE: Routing

Added new routing mechanism to controllers for edit methods.

osmanaliyardim committed:
FEATURE: Caching

Caching mechanism has been applied.

osmanaliyardim committed:
FEAT: Optional Logging

Optional logging based on log level has been applied to open/close logging at any level and anytime.

osmanaliyardim committed:
FIX: Routing

Routing endpoint fixed

osmanaliyardim committed:
REFACTOR: Dto Refactoring

Dtos carried out of data access layer and applied for just MVC layer.

dotnet-fizzyy commented on Business/Concrete/CategoryManager.cs:
```diff
@@ -1,27 +1,72 @@
 
-    public async Task<List<Category>> GetListAsync() => 
-        await _categoryRepository.GetListAsync();
+        FileName = DateTime.Now.ToString("yyyyMMdd") + "-" + FileName?.Trim() + FileExtension;
```
`-` delimiter should be also put in constant to avoid extra questions)

dotnet-fizzyy commented on Business/Concrete/CategoryManager.cs:
```diff
@@ -1,27 +1,72 @@
+        return mappedCategory;
+    } 
+
+    public async Task<List<CategoryDetailDto>> GetListAsync()
```
From business point of view `GetList`, honestly, tells nothing and might be confusing. Let's replace it with `GetAllAsync`

dotnet-fizzyy commented on Business/Concrete/CategoryManager.cs:
```diff
@@ -1,27 +1,72 @@
 
-    public CategoryManager(ICategoryRepository categoryRepository)
+    public CategoryManager(ICategoryRepository categoryRepository,
+        IMapper mapper)
```
This wrap is extra one, let's align arguments on the same line

dotnet-fizzyy commented on Entity/Dtos/ProductDetailDto.cs:
```diff
@@ -1,16 +1,31 @@
 {
-    public string Name { get; set; }
+    public ProductDetailDto(
+    string name,
```
I advise you to add tabulation here to make difference between args and constructor declaration move obvious

dotnet-fizzyy commented on WebNetSample/Controllers/CategoriesController.cs:
```diff
@@ -1,16 +1,20 @@
 
-    public CategoriesController(ICategoryService categoryService)
+    public CategoriesController(ICategoryService categoryService,
+        IMapper mapper)
```
Same case with alignment in one line

osmanaliyardim committed:
REFACTOR: Refactoring Naming

Naming refactorings have been applied.

osmanaliyardim committed:
FEATURE: Dependency Resolver

Dependency resolvers have been applied for both MemoryCache and ICacheService.

osmanaliyardim committed:
REFACTOR: Namings and lambdas

Some code refactorings for namings and lambda usage for code magnification

osmanaliyardim committed:
REFACTOR: Naming, DependencyResolvers

Some code refactorings and dependency resolvers applied for caching mechanism.

osmanaliyardim committed:
REFACTOR: Alignments

Some code aligments for increasing readability.

osmanaliyardim committed:
FEATURE: Routing

Endpoint and routing mechanism for images have been created.

osmanaliyardim committed:
FIX: Logging

Logging dependency resolvers and file location problems solved.

osmanaliyardim committed:
REFACTOR: Tabulation

Tabulation fixed.

osmanaliyardim committed:
Merge branch 'feature/controllers-apps-reqs-task-4' into feature/controllers-apps-reqs-task-5

osmanaliyardim committed:
Merge pull request #10 from osmanaliyardim/feature/controllers-apps-reqs-task-5

FEAT: Optional Logging

osmanaliyardim committed:
Merge branch 'feature/controllers-apps-reqs-task-3' into feature/controllers-apps-reqs-task-4

osmanaliyardim committed:
Merge pull request #9 from osmanaliyardim/feature/controllers-apps-reqs-task-4

FEATURE: Caching

osmanaliyardim committed:
Merge branch 'feature/controllers-apps-reqs-task-2' into feature/controllers-apps-reqs-task-3

osmanaliyardim committed:
Merge pull request #8 from osmanaliyardim/feature/controllers-apps-reqs-task-3

Feature/controllers apps reqs task 3

osmanaliyardim committed:
Merge branch 'dev' into feature/controllers-apps-reqs-task-2
